x/metadata/keeper/specification.go

```diff
@@ -41,9 +93,164 @@ func (k Keeper) GetScopeSpecification(ctx sdk.Context, id types.MetadataAddress)
+	store := ctx.KVStore(k.storeKey)
+
+	// Index all the scope spec owner addresses
+	for _, a := range scopeSpec.OwnerAddresses {
+		addr, err := sdk.AccAddressFromBech32(a)
+		if err == nil {
+			store.Set(types.GetAddressScopeSpecCacheKey(addr, scopeSpec.SpecificationId), []byte{0x01})
+		}
+	}
+
+	// Index all the session spec ids
+	for _, contractSpecID := range scopeSpec.ContractSpecIds {
+		store.Set(types.GetContractSpecScopeSpecCacheKey(contractSpecID, scopeSpec.SpecificationId), []byte{0x01})
+	}
+}
+
+// clearScopeSpecificationIndex removes all indexes for the given scope spec.
+// The provided scope spec must be one that is already stored (as opposed to a new one or updated version of one).
+func (k Keeper) clearScopeSpecificationIndex(ctx sdk.Context, scopeSpec types.ScopeSpecification) {
+	store := ctx.KVStore(k.storeKey)
+
+	// Delete all owner address + scope spec entries
+	for _, a := range scopeSpec.OwnerAddresses {
+		addr, err := sdk.AccAddressFromBech32(a)
+		if err == nil {
+			store.Delete(types.GetAddressScopeSpecCacheKey(addr, scopeSpec.SpecificationId))
+		}
+	}
+
+	// Delete all contract spec + scope spec entries
+	for _, contractSpecID := range scopeSpec.ContractSpecIds {
+		store.Delete(types.GetContractSpecScopeSpecCacheKey(contractSpecID, scopeSpec.SpecificationId))
+	}
+}
+
+// isScopeSpecUsed checks to see if a scope exists that is defined by this scope spec.
+func (k Keeper) isScopeSpecUsed(ctx sdk.Context, id types.MetadataAddress) bool {
+	scopeSpecReferenceFound := false
+	err := k.IterateScopesForScopeSpec(ctx, id, func(scopeID types.MetadataAddress) (stop bool) {
+		scopeSpecReferenceFound = true
+		return true
+	})
+	// If there was an error, that means there was an entry, so return true.
+	return err != nil || scopeSpecReferenceFound
+}
+
+// ValidateScopeSpecUpdate - full validation of a scope specification.
+func (k Keeper) ValidateScopeSpecUpdate(ctx sdk.Context, existing, proposed types.ScopeSpecification, signers []string) error {
+	// IDS must match
+	if len(existing.SpecificationId) > 0 {
+		if !proposed.SpecificationId.Equals(existing.SpecificationId) {
+			return fmt.Errorf("cannot update scope spec identifier. expected %s, got %s",
+				existing.SpecificationId, proposed.SpecificationId)
+		}
+	}
+
+	// Must pass basic validation.
+	if err := proposed.ValidateBasic(); err != nil {
+		return err
+	}
+
+	// Signatures required of all existing data owners.
+	if err := k.ValidateScopeSpecAllOwnersAreSigners(existing, signers); err != nil {
+		return err
+	}
+
+	// Validate the proposed contract spec ids.
+	for _, contractSpecID := range proposed.ContractSpecIds {
+		contractSpec, found := k.GetGroupSpecification(ctx, contractSpecID)
+		// Make sure that all contract spec ids are valid and exist
+		if !found {
+			return fmt.Errorf("no contract spec exists with id %s", contractSpecID)
+		}
+		// Also make sure that the parties in each contract spec are also in the scope spec.
```

Comment:
This relationship goes the other way ... the list of parties required is additive as we go down the tree... From the top for a LOAN only an ORIGINATOR might be required ... but if there is a contract for a loan payment then that contract might require both the ORIGINATOR and the SERVICER for example.